core/logger/audit/audit_logger.go

```diff
@@ -0,0 +1,233 @@
+				return AuditLoggerConfig{}, errors.Errorf("Invalid headers provided for the audit logger. Value, single pair split on || required, got: %s", keyValue)
+			}
+			headers = append(headers, serviceHeader{
+				header: keyValue[0],
+				value:  keyValue[1],
+			})
+		}
+	}
+
+	return AuditLoggerConfig{
+		forwardToUrl:   forwardToUrl,
+		environment:    environment,
+		jsonWrapperKey: jsonWrapperKey,
+		headers:        headers,
+	}, nil
+}
+
+type AuditLoggerService struct {
+	logger          logger.Logger
+	enabled         bool
+	serviceURL      string
+	serviceHeaders  []serviceHeader
+	jsonWrapperKey  string
+	environmentName string
+	hostname        string
+	localIP         string
+	loggingChannel  chan WrappedAuditLog
+}
+
+// Configurable headers to include in POST to log service
+type serviceHeader struct {
+	header string
+	value  string
+}
+
+type WrappedAuditLog struct {
+	eventID EventID
+	data    Data
+}
+
+// NewAuditLogger returns a buffer push system that ingests audit log events and
+// asynchronously pushes them up to an HTTP log service.
+// Parses and validates the AUDIT_LOGS_* environment values and returns an enabled
+// AuditLogger instance. If the environment variables are not set, the logger
+// is disabled and short circuits execution via enabled flag.
+func NewAuditLogger(logger logger.Logger, config *AuditLoggerConfig) (AuditLogger, error) {
+	if config == nil {
+		logger.Info("Audit logger configuration is nil. Cannot start audit logger subsystem and audit events will not be captured.")
+		return &AuditLoggerService{}, nil
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return &AuditLoggerService{}, errors.Errorf("Audit Log initialization error - unable to get hostname: %s", err)
+	}
+
+	loggingChannel := make(chan WrappedAuditLog, bufferCapacity)
+
+	// Finally, create new auditLogger with parameters
+	auditLogger := AuditLoggerService{
+		logger:          logger.Helper(1),
+		enabled:         true,
+		serviceURL:      config.forwardToUrl,
+		serviceHeaders:  config.headers,
+		jsonWrapperKey:  config.jsonWrapperKey,
+		environmentName: config.environment,
+		hostname:        hostname,
+		localIP:         getLocalIP(),
+		loggingChannel:  loggingChannel,
+	}
+
+	// Start our go routine that will receive logs and send them out to the
+	// configured service.
+	go auditLogger.auditLoggerRoutine()
+
+	return &auditLogger, nil
+}
+
+// / Entrypoint for new audit logs. This buffers all logs that come in they will
+// / sent out by the goroutine that was started when the AuditLoggerService was
+// / created. If this service was not enabled, this immeidately returns.
+// /
+// / This function never blocks.
+func (l *AuditLoggerService) Audit(ctx context.Context, eventID EventID, data Data) {
+	if !l.enabled {
+		return
+	}
+
+	wrappedLog := WrappedAuditLog{
+		eventID: eventID,
+		data:    data,
+	}
+
+	select {
+	case l.loggingChannel <- wrappedLog:
+	default:
+		if len(l.loggingChannel) == bufferCapacity {
```

Comment:
This `len()` call is racy. There is no way to ensure that it happens atomically with the failure to send to the channel, and it will incorrectly trigger the else case (possibly quite frequently). By this point the goroutine could have been suspended for an arbitrary amount of time and some space could have cleared in the channel concurrently on another thread/routine, yielding a value less than `cap()/buffercapacity`.

But we already know that `len(l.loggingChannel) == cap(l.loggingChannel) == bufferCapacity` at the time we tried to send, because that is the only case other than `nil` for which the channel send would not be ready:
> The default clauses in the select statements execute when no other case is ready, meaning that the selects never block

https://go.dev/doc/effective_go#leaky_buffer

Reply:
Fixed